worker/worker/parser/category.py

```diff
@@ -16,97 +18,152 @@ def extract_challenge_ids(txt):
-    result = re.findall(pattern, txt)
-    if not result:
-        raise RootMeParsingError("Could not parse challenges.")
-    return result
+def extract_difficulties(r):
+    tree = html.fromstring(r.content)
+    difficulties = tree.xpath('//td[@class="show-for-medium-up"]/a[starts-with(@href,"tag")]/@title')
+    difficulties = list(map(lambda x: x.split(':')[0].strip(), difficulties))
+    return difficulties
+
 
+def extract_values(r):
+    tree = html.fromstring(r.content)
+    values = tree.xpath('//td[@class="show-for-medium-up"]/a[starts-with(@href,'
+                        '"tag")]/parent::td/preceding-sibling::td/text()')
+    values = list(filter(lambda x: '\n' not in x, values))
+    values = list(map(int, values))
+    return values
 
-def extract_info_from_row(row):
-    path, statement, name = extract_challenge_general_info(row)
-    validations_percentage, validations_nb = extract_validation_info(row)
-    value, difficulty = extract_difficulty(row)
-    author = extract_author(row)
-    note, solutions_nb = extract_solutions_and_note(row)
 
-    return {
+def extract_authors(r):
+    tree = html.fromstring(r.content)
+    all_authors = tree.xpath('//td[@class="show-for-large-up"]')
+    #  all_authors = list(map(lambda author: author.getchildren(), all_authors))  #  getchildren is a deprecated method
+    all_authors = [[elements for elements in td] for td in all_authors]  # match "a" elements in td elements
+    authors = []
+    for challenge_authors in all_authors:
+        result = []
+        if challenge_authors:  # challenge_authors != []
+            result = [author.get('href') for author in challenge_authors]
+            result = list(map(lambda x: re.match('\/(.*?)\?', x).group(1), result))
+        authors.append(result)
+    return authors
+
+
+def extract_notes(r):
+    tree = html.fromstring(r.content)
+    notes = tree.xpath('//td/img[starts-with(@src, "squelettes/img/note")]/@src')
+    notes = list(map(lambda x: re.match('.*note(.*?)\.png', x).group(1), notes))
+    notes = list(map(int, notes))
+    return notes
+
+
+def extract_solutions_nbs(r):
+    tree = html.fromstring(r.content)
+    solutions = tree.xpath('//td/img[starts-with(@src, "squelettes/img/note")]/parent::td/following-sibling::td/text()')
+    solutions = list(map(int, solutions))
+    return solutions
+
+
+def extract_challenges_info(r):
+    paths = extract_paths(r)
+    statements = extract_statements(r)
+    names = extract_names(r)
+    validations_percentages = extract_validations_percentages(r)
+    validations_nbs = extract_validations_nbs(r)
+    difficulties = extract_difficulties(r)
+    values = extract_values(r)
+    authors = extract_authors(r)
+    notes = extract_notes(r)
+    solutions_nbs = extract_solutions_nbs(r)
+
+    # TODO: verify that every lists have same lengths
+    response = [{
         'path': path.strip(),
         'statement': unescape(statement).strip(),
         'name': unescape(name).strip(),
```

Comment:
Tu as mis strip() and `extract_author`, pourquoi ne pas mettre strip dans les autres fonctions aussi?
Pareil pour unescape

Reply:
OK